kernel/kernel-api/src/main/java/io/delta/kernel/internal/snapshot/SnapshotManager.java

```diff
@@ -90,6 +95,21 @@ public Snapshot buildLatestSnapshot(TableClient tableClient, Path logPath, Path
         return getSnapshotAtInit(tableClient, logPath, dataPath);
     }
 
+    /**
+     * Updates the current `latestSnapshotHint` with the `newHint` if and only if the newHint is
+     * newer (i.e. has a later table version).
+     * <p>
+     * This doesn't need to be thread-safe. Suppose the current `latestSnapshotHint` is at version
+     * N. If three Snapshots (N-1, N+1, N+2) concurrently register their hints, only N+1 and N+2
+     * will be able to replace the current. And either winning the update is fine.
+     */
+    public void registerHint(SnapshotHint newHint) {
+        if (!latestSnapshotHint.isPresent() ||
+                newHint.getVersion() > latestSnapshotHint.get().getVersion()) {
+            latestSnapshotHint = Optional.of(newHint);
```

Comment:
This code is not thread safe, in spite of the comment stating otherwise. It reads `latestSnapshotHint` twice, and there's no guarantee it stays the same between reads. And then it updates the hint as well, and there's nothing to block the hint from getting changed by somebody else who races in (which could break monotonicity of the hint). Further, the value is not marked volatile, so there's no guarantee that different threads _ever_ see each other's changes.

Reply:
Should probably use an [AtomicReference](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/atomic/AtomicReference.html#get--) instead of `Optional`.